doc/pages/expansions.asciidoc

```diff
@@ -1,199 +1,294 @@
+*%arg{@}*::
+    becomes `$@`
 
-*kak_bufname*::
-    name of the current buffer
+*%opt{x}*::
+    becomes `$kak_opt_x`
 
-*kak_buffile*::
-    full path of the file or same as kak_bufname when there’s no
-    associated file
+*%reg{x}*::
+    (where _x_ is the alphabetic name of a register) +
+    `$kak_reg_x` contains all the selections in register _x_ +
+    `$kak_main_reg_x` contains only the main selection
 
-*kak_buflist*::
-    quoted list of the currently opened buffer names
+*%val{x}*::
+    becomes `$kak_x`
 
-*kak_buf_line_count*::
-    the current buffer line count
+When turned into environment variables, list-type options, `$kak_reg_x`, and
+"quoted list" values will be shell-quoted so the shell doesn't get confused
+about how many items the list contains. You will need to apply `eval` to get
+back the original values. For example, if you want to process the contents
+of each selection, you can do something like:
 
-*kak_timestamp*::
-    timestamp of the current buffer, the timestamp is an integer value
-    which is incremented each time the buffer is modified
+----
+eval set -- $kak_selections
+while [ $# -gt 0 ]; do
+    # ...do a thing with $1...
+    shift
+done
+----
 
-*kak_history_id*::
-    history id of the current buffer, the history id is an integer value
-    which is used to reference a specific buffer version in the undo tree
+The `eval` command will take the expanded `$kak_selections` and unquote them,
+then execute the resulting `set` command, which sets the shell's argument
+variables to the items from `$kak_selections`. The `while` loop with `shift`
+iterates through the arguments one by one.
 
-*kak_runtime*::
-    directory containing the kak support files, determined from kakoune's
-    binary location.
+Only variables actually mentioned in the body of the shell expansion will
+be exported into the shell's environment. For example:
 
-*kak_config*::
-    directory containing the user configuration
+----
+echo %sh{ env | grep ^kak_ }
+----
 
-*kak_version*::
-    version of the current Kakoune server (git hash or release name)
+...will find none of Kakoune's special environment variables, but:
 
-*kak_count*::
-    count parameter passed to the command, defaults to 0 if no count given
+----
+echo %sh{ env | grep ^kak_ # $kak_session }
+----
 
-*kak_register*::
-    register parameter passed to the command
+...will find the `$kak_session` variable because it was mentioned inside the
+shell script, even just in a comment.
 
-*kak_opt_<name>*::
-    value of option *name*
+TIP: These environment variables are also available in other contexts where
+Kakoune uses a shell command, such as the `|`, `!` or `$` normal mode commands
+(See <<keys#,`:doc keys`>>).
 
-*kak_reg_<r>*::
-    quoted list of value of register *r*
+== Value expansions
 
-*kak_main_reg_<r>*::
-    content of register *r* associated with the main selection.
+Expansions with the type `val` can be used in various places, depending on the
```

Comment:
Expansions of type `val` give access to Kakoune internal data that is not stored in an option or a register.

Reply:
Well put, I'll use that.